server_status.py

```diff
@@ -0,0 +1,168 @@
+import subprocess
+import os
+import time
+import mraa
+import inspect
+import sys
+import math
+
+os.chdir('/home/root/gpio')
+print "*** START server_status.py ***"
+
+## INITIALISE VARIABLES
+
+red = mraa.Gpio(31)     # 'GP44'
+red.dir(mraa.DIR_OUT)
+green = mraa.Gpio(45)   # 'GP45'
+green.dir(mraa.DIR_OUT)
+blue = mraa.Gpio(32)    # 'GP46'
+blue.dir(mraa.DIR_OUT)
+
+buendia = 0       # localhost status
+openmrs_ext = 0   # openmrs server status
+openmrs_int = -1  # openmrs internal status
+buendia_url = 'localhost'
+openmrs_url = 'localhost:9000/openmrs/index.htm'
+
+last_check = time.time()
+last_log_trunc = time.time()
+
+# reset openmrs internal status to match openmrs_int
+subprocess.call('echo -1 > /home/root/debian/home/buendia/server_status.txt', shell=True)
+
+
+## FUNCTIONS
+
+# Returns next line number (debugging)
+def report():
+  line = inspect.currentframe().f_back.f_lineno
+  print time.strftime('%X') + ' - line ' + str(line + 1)
+
+def reset():
+  red.write(0)
+  green.write(0)
+  blue.write(0)
+
+# flash list of LEDs for individual periods and total duration (cycles rounded to multiple of len(leds) )
+def flash_colours(leds, duration, period):
+  n = len(leds)
+  for i in range(int( n * math.floor(.5 + duration/(n*n*period)) )):
+    leds[i%n].write(1)
+    time.sleep(period)
+    leds[i%n].write(0)
+    if n == 1:
+      time.sleep(period)
+
+def openmrs_internal_status(): # runs for ~1 second regardless of state
+  global openmrs_int
+  with open('/home/root/debian/home/buendia/server_status.txt', "r") as status_file:
+    openmrs_int_new = int(status_file.read())
+  if not openmrs_int_new == openmrs_int:  # state has changed
+    print time.strftime('%X') + ' - openmrs_int change: ' + str(openmrs_int) + ' > ' + str(openmrs_int_new)
+    reset()
+    openmrs_int = openmrs_int_new
+  # TASK LED's according to OpenMRS server status
+  if openmrs_int == 0:       # down - flash RED
+    flash_colours([red], 1, .5)
+  elif openmrs_int == 1:     # normal use - BLUE on
+    blue.write(1)
+    time.sleep(1)
+  elif openmrs_int == 2:     # activity  - blink BLUE on/off
+    flash_colours([blue], 1, .05)
+  elif openmrs_int == 3:     # backing-up - blink BLUE/GREEN
+    flash_colours([blue, green], 1, .1)
+  elif openmrs_int == 4:     # back-up failed - blink BLUE/RED
+    flash_colours([blue, red], 1, .5)
+  else:
+    time.sleep(1)
+
+# classify server repsonses - i.e. interpret top line of 'curl -Is <url>'
+def check_url(url):
+  global last_check
+  response = subprocess.check_output("curl --max-time 1 -Is " + url + " | head -n 1", shell=True)
```

Comment:
Have a look at the buendia-server and buendia-dashboard packages:
  - buendia-server runs buendia-warmup once a minute
  - buendia-dashboard shows the last result from buendia-warmup as the
up/down status of the server

Reuse the output of buendia-warmup (/var/log/buendia/buendia-warmup.log) or
reuse buendia-warmup (tools/warmup) instead of these curl calls.  (There's
even a check_url function in there.)



On Tue, Nov 10, 2015 at 2:31 PM, Robin Edwards <notifications@github.com>
wrote:

> In server_status.py
> <https://github.com/projectbuendia/server-status/pull/2#discussion_r44475695>
> :
>
> > +  elif openmrs_int == 1:     # normal use - BLUE on
> > +    blue.write(1)
> > +    time.sleep(1)
> > +  elif openmrs_int == 2:     # activity  - blink BLUE on/off
> > +    flash_colours([blue], 1, .05)
> > +  elif openmrs_int == 3:     # backing-up - blink BLUE/GREEN
> > +    flash_colours([blue, green], 1, .1)
> > +  elif openmrs_int == 4:     # back-up failed - blink BLUE/RED
> > +    flash_colours([blue, red], 1, .5)
> > +  else:
> > +    time.sleep(1)
> > +
> > +# classify server repsonses - i.e. interpret top line of 'curl -Is <url>'
> > +def check_url(url):
> > +  global last_check
> > +  response = subprocess.check_output("curl --max-time 1 -Is " + url + " | head -n 1", shell=True)
>
> It's not replicating the process. Intention is to integrate the two - it's
> just not yet clear how. Dan can comment on broader decision to add LED
> status reporting..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/projectbuendia/server-status/pull/2/files#r44475695>.
>